src/snyk/common/services/learnService.ts

```diff
@@ -0,0 +1,115 @@
+import axios from 'axios';
+import { isCodeIssue, isOssIssue, OpenCommandIssueType } from '../../common/commands/types';
+import type { completeFileSuggestionType } from '../../snykCode/interfaces';
+import { OssIssueCommandArg } from '../../snykOss/views/ossVulnerabilityTreeProvider';
+import { ILog } from '../logger/interfaces';
+
+const ecosystems = [
+  'all',
+  'java',
+  'javascript',
+  'python',
+  'golang',
+  'php',
+  'cpp',
+  'csharp',
+  'kubernetes',
+  'ruby',
+  'elixir',
+  'docker',
+  'npm',
+] as const;
+type LessonEcosystem = typeof ecosystems[number];
+
+const isValidEcosystem = (potentialEcosystem: string | null | undefined): potentialEcosystem is LessonEcosystem => {
+  return ecosystems.includes(potentialEcosystem as LessonEcosystem);
+};
+
+export type Lesson = {
+  title: string;
+  lessonId: string;
+  ecosystem: LessonEcosystem;
+  url: string;
+};
+
+type PackageMapper = {
+  [key in LessonEcosystem]?: string[];
+};
+
+export interface ILearnService {
+  getLesson(options: { ecosystem: LessonEcosystem; cwes?: string[]; rules?: string[] }): Promise<Lesson | null>;
+}
+
+export class LearnService implements ILearnService {
+  private ecosystem: LessonEcosystem;
+  private cwes: string[] = [];
+
+  constructor(
```

Comment:
Let's have this class as singleton and pass it through to webviewprovider constructors. Webview provider will call a method on it to pull the lesson, where all current constructor orchestration would take place.

Reply:
Hmm, the class OssSuggestionWebviewProvider doesn't currently have awareness of the vulnerability object in the constructor. I might not be understanding your suggestion. I'll message you

Reply:
Outcome:
- Refactor the Learn service so that it does not hold any vulnerability level state. Only:
1. Caching
2. Knowing how to take an issue and make a request
3. Knowing whether a learn link should ever be shown
- Create an instance of the learn service in the `BaseSnykModule` constructor
- Pass this instance to the `OssSuggestionWebviewProvider` and Code... constructors
- The WebviewProviders can directly use this instance